cnd/src/logging/initialize.rs

```diff
@@ -0,0 +1,214 @@
+use fern::{Dispatch, FormatCallback};
+use log::{LevelFilter, Record};
+use std::{fmt::Arguments, io::stdout};
+
+pub fn initialize(
+    base_log_level: LevelFilter,
+    structured: bool,
+) -> Result<(), log::SetLoggerError> {
+    #![allow(clippy::print_stdout)] // We cannot use `log` before we have the config file
+    println!("Initializing logging with base level {}", base_log_level);
+
+    let (max_level, log) = create_logger(base_log_level, structured, stdout());
+
+    log::set_boxed_logger(log)?;
+    log::set_max_level(max_level);
+
+    Ok(())
+}
+
+fn create_logger<T: Into<fern::Output>>(
+    base_log_level: LevelFilter,
+    structured: bool,
+    target: T,
+) -> (LevelFilter, Box<dyn log::Log>) {
+    let formatter = if structured {
+        json_formatter
+    } else {
+        line_formatter
+    };
+
+    Dispatch::new()
+        .format(formatter)
+        .level(base_log_level)
+        .level_for("tokio_core::reactor", LevelFilter::Info)
+        .level_for("tokio_reactor", LevelFilter::Info)
+        .level_for("hyper", LevelFilter::Info)
+        .level_for("warp", LevelFilter::Info)
+        .level_for("libp2p", LevelFilter::Debug) // the libp2p library
+        .level_for("sub-libp2p", LevelFilter::Debug) // the libp2p subsystem in our application
+        .level_for("http-api", LevelFilter::Debug) // the http-api of our application
+        .chain(target)
+        .into_log()
+}
+
+fn line_formatter(out: FormatCallback<'_>, message: &Arguments<'_>, record: &Record<'_>) {
+    out.finish(format_args!(
+        "[{date}][{level}][{path}{line}] {message}",
+        date = chrono::Local::now().format("%Y-%m-%d %H:%M:%S%.3f"),
+        path = record.target(),
+        line = record
+            .line()
+            .map(|line| format!(":{}", line))
+            .unwrap_or_else(String::new),
+        level = record.level(),
+        message = message,
+    ))
+}
+
+/// Formats each log record as a JSON statement
+///
+/// The implementation of this function on purpose does not use `serde_json` to
+/// create the JSON string. Here is why:
+/// 1. `out.finish` only accepts `std::fmt::Arguments` and hence we can only use
+/// `format_args!`
+///
+/// 2. (1) is very likely due to the fact that we don't want to
+/// unnecessarily allocate strings in this function for calls to `Log::log` with
+/// a level that is not even active (think about all the calls to `log::trace`
+/// in all our depenencies that we don't see)
+///
+/// Note that this formatter **CANNOT** handle multiline message as for example
+/// produced by: `log::info!("{:#?}", my_struct);`
```

Comment:
So what's the plan? How do we handle this scenario?

Reply:
I'd propose to just not use these log statements :)

Reply:
Does the compilation fail if we were to use them?